istft: require complex input

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86628
* #86724

Real dtype input to `torch.istft` has been deprecated since PyTorch
1.8, so it is more than passed its due date to be removed.

BC-breaking message:

`torch.istft` no longer supports input in the form of real tensors
with shape `(..., 2)` to mimic complex tensors. Instead, convert
inputs to a complex tensor first before calling `torch.istft`.

1.13
```python
>>> t = torch.rand(65, 33, 2)
>>> _ = torch.istft(t, n_fft=128, length=1024)
```

2.0
```python
>>> t = torch.rand(65, 33, 2)
>>> t_complex = torch.view_as_complex(t)
>>> _ = torch.istft(t_complex, n_fft=128, length=1024)

peterbell10 committed:
stft: move towards always returning complex

For `stft` this makes all cases where `return_complex` default to
`False` into an error and adds a warning when `return_complex=False`
is passed explicitly.

For `istft` this raises an error if the input is not a complex tensor.

[ghstack-poisoned]

peterbell10 committed:
Update on "istft: require complex input"


Real dtype input to `torch.istft` has been deprecated since PyTorch
1.8, so it is more than passed its due date to be removed.

BC-breaking note:
`torch.istft` no longer supports input in the form of real tensors
with shape `(..., 2)` to mimic complex tensors. Instead, convert
inputs to a complex tensor first before calling `torch.istft`.

[ghstack-poisoned]

peterbell10 committed:
Update on "istft: require complex input"


Real dtype input to `torch.istft` has been deprecated since PyTorch
1.8, so it is more than passed its due date to be removed.

BC-breaking message:

`torch.istft` no longer supports input in the form of real tensors
with shape `(..., 2)` to mimic complex tensors. Instead, convert
inputs to a complex tensor first before calling `torch.istft`.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "istft: require complex input"


Real dtype input to `torch.istft` has been deprecated since PyTorch
1.8, so it is more than passed its due date to be removed.

BC-breaking message:

`torch.istft` no longer supports input in the form of real tensors
with shape `(..., 2)` to mimic complex tensors. Instead, convert
inputs to a complex tensor first before calling `torch.istft`.

[ghstack-poisoned]

mruberry commented on torch/functional.py:
```diff
@@ -678,14 +678,13 @@ def stft(input: Tensor, n_fft: int, hop_length: Optional[int] = None,
+        where the ``channel`` dimension is optional.
+
+        .. versionchanged:: 1.14.0
+            Real input is no longer supported. Input must be complex, as returned by
```
"Real-valued inputs are no longer supported. Input must now have a complex datatype, ..."

mruberry reviewed (approved):
Cool! A couple small nits inline for your review

peterbell10 committed:
Update on "istft: require complex input"


Real dtype input to `torch.istft` has been deprecated since PyTorch
1.8, so it is more than passed its due date to be removed.

BC-breaking message:

`torch.istft` no longer supports input in the form of real tensors
with shape `(..., 2)` to mimic complex tensors. Instead, convert
inputs to a complex tensor first before calling `torch.istft`.

[ghstack-poisoned]
